### PR TITLE
Text on datagrams

### LINKED ISSUE #269: Mention the datagram extension?
So far we don't mention any extension in the ops drafts, however, datagrams actually require a different interface and as such are important for an application to know about, therefore, I think it would be good to add a separate section in the applicability statement and mention it as a separate service over QUIC.

### PR DESCRIPTION
**Pull request by mirjak**
resolves #269

### CODE REVIEW

File: draft-ietf-quic-applicability.md [hunk 1]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
 old version, stop sending the old version in Version Negotiation packets, then
 the old version is no longer accepted.
 
+# Unreliable Datagram Service over QUIC
+
+{{?I-D.draft-ietf-quic-datagram}} specifies an QUIC extension to enable sending
+and receiving unreliable datagrams over QUIC. Other than with the use of UDP
+only, the datagram service in QUIC is congestion-controlled, and thereby
```
Comment (ianswett) [inline draft-ietf-quic-applicability.md:716]: What does "Other than with the use of UDP only, the datagram service in QUIC is congestion-controlled," mean?

File: draft-ietf-quic-applicability.md [hunk 2]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+supporting application designers to address the requirements in {{?RFC8085}}.
+QUIC datagrams are not flow-controlled and data chunks may get drop if the
+receiver is overloaded.
```
Comment (ianswett) [inline draft-ietf-quic-applicability.md:719]: I'd move this last sentence into the next paragraph, because lack of flow control is just another way in which Datagrams are not reliable.

File: draft-ietf-quic-applicability.md [hunk 3]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+
+The reliable transmission service of QUIC usually provides a stream-based
+interface to send and received data over eventually multiple QUIC streams.
```
Comment (ianswett) [inline draft-ietf-quic-applicability.md:722]:
```suggestion
The reliable transmission service of QUIC provides a stream-based
interface to send and receive data over multiple QUIC streams.
```

File: draft-ietf-quic-applicability.md [hunk 4]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+The datagram service, however, usually provides a message-based interface,
```
Comment (ianswett) [inline draft-ietf-quic-applicability.md:723]:
```suggestion
The datagram service, however, provides a message-based interface,
```

Unless there's a reason to use usually and you're going to explain what happens when that's not the case, I think it's clearer without.

File: draft-ietf-quic-applicability.md [hunk 5]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+with the expectation that an application layer framing exists on top to
+recognise messages boundaries as well as the order of received message chunks
+if needed.
```
Comment (ianswett) [inline draft-ietf-quic-applicability.md:726]:
```suggestion
Application layer framing can exist on top to recognise messages boundaries
as well as the order of received message chunks if needed.
```

File: draft-ietf-quic-applicability.md (hunk 1 above)
Comment (tfpauly) [reply draft-ietf-quic-applicability.md:716]: This probably would be cleared if it were phrased as being about unreliable datagrams, not UDP. Tunneling UDP is just one way to use datagram frames.

Comment (martinduke) [reply draft-ietf-quic-applicability.md:716]: I would rephrase this as "Unlike operating directly over UDP, an applications that uses the QUIC datagram service need not provide its own congestion control, per {{?RFC8085}}, as QUIC datagrams are congestion controlled."

**Review by martinduke — CHANGES_REQUESTED** [review]

File: draft-ietf-quic-applicability.md (hunk 1 above)
Comment (mirjak) [reply draft-ietf-quic-applicability.md:716]:
```suggestion
and receiving unreliable datagrams over QUIC. Unlike operating directly over UDP,
an applications that uses the QUIC datagram service need not provide its own
congestion control, per {{?RFC8085}}, as QUIC datagrams are congestion controlled.
```

File: draft-ietf-quic-applicability.md [hunk 6]
```diff
@@ -709,6 +709,22 @@ new version uses the same process in reverse.  Servers disable validation of the
 old version, stop sending the old version in Version Negotiation packets, then
 the old version is no longer accepted.
 
+# Unreliable Datagram Service over QUIC
+
+{{?I-D.draft-ietf-quic-datagram}} specifies an QUIC extension to enable sending
+and receiving unreliable datagrams over QUIC. Other than with the use of UDP
+only, the datagram service in QUIC is congestion-controlled, and thereby
+supporting application designers to address the requirements in {{?RFC8085}}.
```
Comment (mirjak) [inline draft-ietf-quic-applicability.md:717]:
```suggestion
```

File: draft-ietf-quic-applicability.md (hunk 4 above)
Comment (mirjak) [reply draft-ietf-quic-applicability.md:723]: Usually is there because you can also implement an message base interface for stream (see taps) and there is nothing in the protocol spec that says anything about the interface.

File: draft-ietf-quic-applicability.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-quic-applicability.md:716]: Yes, thanks. That seems clearer to me, except s/its/their

**Comment by mirjak** [comment]
I updated the text of the second paragraph. If people could have another look that would be welcome. But I believe we are ready to merge (after the meeting).

File: draft-ietf-quic-applicability.md [hunk 7]
```diff
@@ -709,6 +709,21 @@ new version uses the same process in reverse.  Servers disable validation of the
 old version, stop sending the old version in Version Negotiation packets, then
 the old version is no longer accepted.
 
+# Unreliable Datagram Service over QUIC
+
+{{?I-D.draft-ietf-quic-datagram}} specifies an QUIC extension to enable sending
```
Comment (tfpauly) [inline draft-ietf-quic-applicability.md:714]:
```suggestion
{{?I-D.draft-ietf-quic-datagram}} specifies a QUIC extension to enable sending
```

File: draft-ietf-quic-applicability.md [hunk 8]
```diff
@@ -709,6 +709,21 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+and receiving unreliable datagrams over QUIC. Unlike operating directly over
+UDP, applications that use the QUIC datagram service do not need to implement
+their own congestion control, per {{?RFC8085}}, as QUIC datagrams are
+congestion controlled.
+
+QUIC datagrams are not flow-controlled and as such data chunks may get drop
```
Comment (tfpauly) [inline draft-ietf-quic-applicability.md:720]:
```suggestion
QUIC datagrams are not flow-controlled, and as such data chunks may be dropped
```

File: draft-ietf-quic-applicability.md [hunk 9]
```diff
@@ -709,6 +709,21 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+if the receiver is overloaded. While the reliable transmission service of QUIC
+provides a stream-based interface to send and receive data in order over
+multiple QUIC streams, the datagram service has a unordered message-based
+interface. If needed, an application layer framing should be used on top to
```
Comment (tfpauly) [inline draft-ietf-quic-applicability.md:724]:
```suggestion
interface. If needed, an application layer framing can be used on top to
```

File: draft-ietf-quic-applicability.md [hunk 10]
```diff
@@ -709,6 +709,21 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+recognise the order of received message chunks.
```
Comment (tfpauly) [inline draft-ietf-quic-applicability.md:725]: I think it's more about demultiplexing than ordering, since ordering is implicit in the QUIC packet numbers.

```suggestion
allow separate flows of unreliable datagrams to be multiplexed on one QUIC connection.
```

**Review by tfpauly — CHANGES_REQUESTED** [review]

File: draft-ietf-quic-applicability.md [hunk 11]
```diff
@@ -709,6 +709,21 @@ new version uses the same process in reverse.  Servers disable validation of the
...
+UDP, applications that use the QUIC datagram service do not need to implement
+their own congestion control, per {{?RFC8085}}, as QUIC datagrams are
+congestion controlled.
+
+QUIC datagrams are not flow-controlled, and as such data chunks may be dropped
+if the receiver is overloaded. While the reliable transmission service of QUIC
+provides a stream-based interface to send and receive data in order over
+multiple QUIC streams, the datagram service has a unordered message-based
+interface. If needed, an application layer framing can be used on top to
+allow separate flows of unreliable datagrams to be multiplexed on one QUIC connection.
```
Comment (mirjak) [inline draft-ietf-quic-applicability.md:725]:
```suggestion
allow separate flows of unreliable datagrams to be multiplexed on one QUIC
connection.
```

**Review by tfpauly — APPROVED** [review]
